File: Solutions/Cpp/container_with_most_water.cpp
/*
    [MEDIUM]
    11. Container With Most Water

    Concepts:
    - array
    - two pointers

    Stats:
        Runtime | 3 ms      [Beats 35.68%]
        Memory  | 62.87 MB  [Beats 78.85%]
*/

class Solution {
public:
    int maxArea(vector<int>& height) {
        int max_water = 0;
        int n = height.size();
        int left = 0, right = n-1;

        while (left < right) {
            max_water = std::max(
                max_water,
                std::min(
                    height[left],
                    height[right]
                ) * (right - left)
            );

            if (height[left] > height[right])
                right--;
            else
                left++;
        }

        return max_water;
    }
};

File: Solutions/Rust/container_with_most_water.rs
/*
    [MEDIUM]
    11. Container With Most Water

    Concepts:
    - array
    - two pointers

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 2.88 MB   [Beats 99.58%]
*/
use std::cmp::{min, max};

impl Solution {
    pub fn max_area(mut height: Vec<i32>) -> i32 {
        let mut max_water: usize = 0;
        let n: usize = height.len();
        let [mut left, mut right]: [usize; 2] = [0, n-1];

        while left < right {
            max_water = max(
                max_water,
                min(
                    height[left],
                    height[right]
                ) as usize * (right - left)
            );

            if height[left] > height[right] { right -= 1; }
            else { left += 1; }
        }

        return max_water as i32;
    }
}